### PR TITLE
Patient centric view support

### PR DESCRIPTION
**Pull request by SCdF**
Migration to create patient person documents from registrations if they don't already exist.

Fixes medic/medic-webapp#3115, medic/medic-webapp#3075 and partially fixes medic/medic-webapp#3082.

### CODE REVIEW

File: migrations/create-patient-contacts.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+          if (err) {
+            return callback(err);
+          }
+
+          var errors = results.filter(function(result) {
+            return !result.ok;
+          });
+
+          if (errors.length) {
+            return callback(new Error('Bulk create errors: ' + JSON.stringify(errors)));
```
Comment (garethbowen) [inline migrations/create-patient-contacts.js]: Very nice!

File: migrations/create-patient-contacts.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+          }
+
+          console.log('DONE');
```
Comment (garethbowen) [inline migrations/create-patient-contacts.js]: Probably unnecessary and maybe misleading. Could replace with "batch done" or just remove.

File: migrations/create-patient-contacts.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      var existingContactShortcodes = _.pluck(results.rows, 'key');
+
+      callback(batch.filter(function(row) {
+        return !_.contains(existingContactShortcodes, row[0]);
+      }));
+    });
+};
+
+var batchCreatePatientContacts = function(batch, callback) {
+  process.stdout.write('Of ' + batch.length + ' potential patients ');
```
Comment (SCdF) [inline migrations/create-patient-contacts.js]: It's a slowly building sentence. This is how you do it to keep it on one line: console.log has a forced newline at the end.

File: migrations/create-patient-contacts.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  filterThoseWithExistingContacts(batch, function(filteredBatch) {
```
Comment (SCdF) [inline migrations/create-patient-contacts.js]: Whoops, nice catch

File: migrations/create-patient-contacts.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    process.stdout.write(filteredBatch.length + ' do not have a contact. ');
+    if (filteredBatch.length === 0) {
+      process.stdout.write('\n');
+      return callback();
+    }
+
+    process.stdout.write('Getting registrations.. ');
+
+    var registrationIdsToConsider = _.flatten(_.pluck(filteredBatch, 1));
```
Comment (SCdF) [inline migrations/create-patient-contacts.js]: Okey dokey

Comment (SCdF) [inline migrations/create-patient-contacts.js]: Oh wait the reason I didn't do that is that it was strictly a filter i.e `A: Coll<T>, P: <Predicate> -> B: Coll<T> where B.length <= A.length` or however you want to represent that. I'll change it and maybe consider changing the function name as well.

File: migrations/create-patient-contacts.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        .uniq(false, function(registration) {
+          // And we only need one for each patient.
+          return registration.patient_id;
+        })
+        .value();
+
+      if (!uniqueValidRegistrations.length) {
+        console.log('no new patient registrations in this batch');
+        return callback();
+      } else {
```
Comment (SCdF) [inline migrations/create-patient-contacts.js]: Okey dokey

File: migrations/create-patient-contacts.js (hunk 2 above)
Comment (SCdF) [inline migrations/create-patient-contacts.js]: I'll change it to `batch done`. It's concluding the one line of output per batch in the log

File: migrations/create-patient-contacts.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        }, {})
+      );
+
+      console.log('There are ' +
+                  registrationsForPatientShortcode.length +
+                  ' patients with registrations');
+
+      async.doWhilst(
+        function(callback) {
+          var batch = registrationsForPatientShortcode.splice(0, BATCH_SIZE);
```
Comment (SCdF) [inline migrations/create-patient-contacts.js]: I'll look into it

Comment (SCdF) [inline migrations/create-patient-contacts.js]: OK output (with a temp batch size of 10) looks like this:
```
Running migration "create-patient-contacts"...
There are 84 patients with registrations
[10/84] Of 10 potential patients 24 do not have a contact. Getting registrations.. no new patient registrations in this batch
[20/84] Of 10 potential patients 1 do not have a contact. Getting registrations.. no new patient registrations in this batch
[30/84] Of 10 potential patients 1 do not have a contact. Getting registrations.. no new patient registrations in this batch
[40/84] Of 10 potential patients 0 do not have a contact.
[50/84] Of 10 potential patients 9 do not have a contact. Getting registrations.. no new patient registrations in this batch
[60/84] Of 10 potential patients 0 do not have a contact.
[70/84] Of 10 potential patients 2 do not have a contact. Getting registrations.. no new patient registrations in this batch
[80/84] Of 10 potential patients 1 do not have a contact. Getting registrations.. no new patient registrations in this batch
[90/84] Of 4 potential patients 0 do not have a contact.
Migration "create-patient-contacts" completed successfully
```

**Comment by SCdF** [comment]
Yeah it is kind of dense, for sure. I think I love `_.chain` too much. Clojure changed me. Would be interested to see different takes on this. Your ease in reading is more important than my ease in writing.

File: migrations/create-patient-contacts.js (hunk 3 above)
Comment (garethbowen) [inline migrations/create-patient-contacts.js]: TIL!

**Comment by mandric** [comment]
@SCdF One thing we'll need to know here is how we will handle potential collisions here. Will the migration just die?  Will the patient records get merged or force update a rev?  I don't think it matters what we do as much as we just know what will happen.

**Comment by SCdF** [comment]
@mandric what do u mean? what would an example be?

**Comment by mandric** [comment]
If we have two records with the same value for `patient_id` (pre migration) what happens/how does the migration handle that?

**Comment by SCdF** [comment]
The migration specifically looks for registrations that create ids, so they should always be unique. I.e. we are looking for forms that represent SMS conversations like:
```
> P Sally 2
< Thanks, Sally has been registered as 12345
```

And then creating a contact with `patient_id: 12345`.

**Pull request merged**